### ISSUE TITLE
[BUG] `InterfaceStubGenerator` Utilizes 10-20% of CPU with Simple Keypresses

### ISSUE THREAD

**Issue by Mike-E-angelo** [reactiveui/refit#1199]
**Describe the bug**
This is in reference to this reported issue here:

https://developercommunity.visualstudio.com/t/ServiceHubRoslynCodeAnalysisServiceSe/1441386

Basic keypresses in a `.cs` file will cause `ServiceHub.RoslynCodeAnalysisService.exe` to utilize 10-20% of available CPU.

Upon further investigation, the issue does not present itself until a Blazor Server-side project is added to the solution.  For your reference and convenience, I have provided an SLN that demonstrates the issue here:

https://github.com/DragonSpark/Framework/blob/issues/reported/refit/The%20DragonSpark%20Framework.sln

**Steps To Reproduce**

1. Open [this file](https://github.com/DragonSpark/Framework/blob/issues/reported/refit/DragonSpark.Application/Security/Identity/Profile/UserSynchronization.cs#L31), placing your cursor at the end of the highlighted line.
2. Press `Enter` a few times.
3. Press `Backspace` a few times.
4. Repeat steps `2` and `3` a few times.
5. Observe CPU utilization by `ServiceHub.RoslynCodeAnalysisService.exe` to be between 10-15%.

**Expected behavior**

Basic key presses in a `.cs` file do not consume so much CPU. :)

**Screenshots**

As this is not a user interface issue, I am not sure this applies.  However, if needed, screenshots demonstrating system taxation can be found in the above reported issue on developercommunity.

**Environment**

- OS:       Windows 10 
- Device:   Desktop 
- Version:   6.0.38 
- Working Version: 6.0.38

**Additional context**

Note that I am using `Microsoft Visual Studio Community 2019 Preview Version 16.11.0 Preview 3.0`.

Please let me know if there is any further information I can provide to further help diagnose this issue.  Outside of this, Refit is an awesome product.  I am a big fan. 👍

**Comment by clairernovotny** [comment]
This seems to be an issue with code generators in general. There's nothing Refit can do here right now.

**Comment by Mike-E-angelo** [comment]
Thank you for the reply, @clairernovotny.  Are you able to provide some more context around this, by chance?  I would like to return this back to the originating issue and see if we can get the correct team aware of this if possible.

**Comment by Mike-E-angelo** [comment]
FWIW running a quick dotTrace session on `ServiceHub.RoslynCodeAnalysisService.exe` with the above steps does show `Refit.Generator` taking most of the time:

![](https://i.imgur.com/fY2CbrB.png)

**Comment by Mike-E-angelo** [comment]
Digging further into this. 😁  It would seem that [these lines are the primary culprits](https://github.com/reactiveui/refit/blob/main/InterfaceStubGenerator.Core/InterfaceStubGenerator.cs#L59-L60).

I am wondering if we [should modify this line](https://github.com/reactiveui/refit/blob/main/InterfaceStubGenerator.Core/InterfaceStubGenerator.cs#L47) to check if either `receiver.CandidateMethods` or `receiver.CandidateInterfaces` have elements in them (and return if not).

Those properties are populated with the latest syntax changes, correct?  If no candidate methods/interfaces are in the latest syntax changes, then `GenerateInterfaceStubs` should not continue.

I may be mistaken here as I am not too entirely familiar with source generators.  My primary motivation is to keep my PC from being my personal furnace. Especially in the summer. :)  

If I am not mistaken, please let me know and I would be interested in providing a PR for a review/fix. 👍

**Comment by Mike-E-angelo** [comment]
> If I am not mistaken

Turns out that I was indeed mistaken.

I have a better understanding on what you mean now, @clairernovotny.  I attempted to put in my suggested guard around interfaces/methods, but that didn't do a dang thing.  It would seem that when a syntax notification occurs, the entirety of the solution/project tree is loaded, rather than the current file under edit.  Subsequently, when `Microsoft.CodeAnalysis.Compilation.GetTypeByMetadataName` is called, it has to compile/process that entire tree.  For larger solutions such as the one I am slaving in, this takes more time to process, hence the CPU churn.

I have opened an issue in the Rosyln repository here:
https://github.com/dotnet/roslyn/issues/54914

I will close this for now but may re-open if it is further determined that Refit can further improve its implementation.  Thank you for your patience as I have taken the time to better understand the different pieces involved here.

**Comment by Mike-E-angelo** [comment]
Looks like the Roslyn team isn't too convinced that this issue is in their court, so until a definitive resolution is found between the two repositories I will keep this issue opened.

**Comment by Mike-E-angelo** [comment]
I am not too familiar with source generators (although getting a bit of a crash course here).  Is there not a way to make this source generator only fire during compilation rather than on keypresses?  Is there a reasoning behind syntax notifications vs. compile?

**Comment by RikkiGibson** [comment]
Source generators are pretty much designed to run on keystroke in the IDE. The V2 source generators API makes the performance of that a lot better because it has the generator register what specific inputs it needs to produce specific outputs and the language service can avoid calling it except when those inputs have changed.

**Comment by Mike-E-angelo** [comment]
Thank you for the response and additional context @RikkiGibson.  I am still learning about this, however, so please bear with me.  From all the research that I have done with source generators, I do not see anything about per-key-press invocation but rather it would seem that the expectation is that source generators execute during compile time.  Even @jaredpar's expectation seemed to [lean towards compile-time execution](https://github.com/dotnet/roslyn/issues/54914#issuecomment-883663981) when initially examining this issue.

Documentation around this seems sparse since this is a new technology, most of it that I found being in the preview/beta stage.  I have yet to find any resource that actually says where source generators are configured to be per-key-press.  If you have a resource it would be greatly appreciated.  

My general feeling for this issue ATM (outside of waiting for v2 and subsequent PR) is to see if we can mitigate it by aligning it with the prevailing documentation out there and only execute at compile time rather than per-key-press.

I will further explore this and update here accordingly.

**Comment by Mike-E-angelo** [comment]
[I have found my answer](https://github.com/dotnet/roslyn/issues/54914#issuecomment-884018143).

**Comment by Mike-E-angelo** [comment]
I wanted to update this issue as it is still very much alive in my world.

It turns out this was _way_ more involved than I initially perceived, like much of software.  I have learned a _lot_ investigating it and have a better understanding of the inner workings of source generation, but more importantly `ServiceHub.RoslynCodeAnalysisService.exe`.

Based on my experience with Visual Studio 2022 Preview 3.1 and the latest .NET6 RC1 SDK, when I first open a solution, source generation works wonderfully.  Refit is a champ:

![](https://i.imgur.com/Vsop3vG.png)

However, over time, a certain something get triggered in `ServiceHub.RoslynCodeAnalysisService.exe` and leads to the following:

![](https://i.imgur.com/sVSRJij.png)

As mentioned, now that we're in `net6.0` here, `JsonSourceGenerator` is taken along for the ride, too.

So, TLDR :P this is definitely not a Refit issue but something that is occurring with `ServiceHub.RoslynCodeAnalysisService.exe` and all source generators after many hours of development.

This has been reported here:
https://developercommunity.visualstudio.com/t/Roslyn-Degrades-over-Time-Placing-Press/1508373
